typing/typecore.ml

```diff
@@ -1812,55 +1812,28 @@ and type_pat_aux
         pat_type = instance expected_ty;
         pat_attributes = sp.ppat_attributes;
         pat_env = !env })
-  | Ppat_or(sp1, sp2) ->
-      let may_split, must_split =
+  | Ppat_or(sp1, sp2) when mode <> Normal ->
+      (* We are in counter-example mode, but try to avoid backtracking *)
```

Comment:
I do not understand your problem, just below that we call `get_splitting_mode`, and have an `assert false` if this was not the expected case.
Would you be happy if I write `when get_splitting_mode mode <> None`?
I would prefer keeping the two cases at the level of the main pattern-matching, and not add an extra parameter to all the other cases, so I don't see how to do it in another way.
(Actually, if we were following coq-style indentation this would be less of a problem.)

Reply:
```ocaml
| Ppat_or(sp1, sp2) ->
  begin match mode with
    ...
  end
```

P.S.: `when` clauses are generally a code smell. They make the code worse, for example here.

Reply:
Indeed, by mixing Coq indentation with ocaml indentation one can avoid increasing the indentation.
Since I don't like to play with indentation I just wrote the way it should.
I preferred the first version, since it emphasized that there were two separate cases, but I have no time to discuss that, so so be it.

Reply:
I hope that it's clear that my point was about doing a proper pattern-match on `mode` instead of assuming things from `mode <> Normal`, and not about the indentation. I don't see how you can prefer a version that has strictly more `assert false` in the code, but at least the new version is nice, thanks!